src/main/java/org/breskul/pool/PooledDataSource.java

```diff
@@ -0,0 +1,129 @@
+        return connectionPool.poll();
+    }
+
+    @Override
+    public Connection getConnection(String username, String password) {
+        return this.getConnection();
+    }
+
+    public String getDriverName() {
+        try {
+            return DriverManager.getDriver(this.url).toString();
+        } catch (SQLException e) {
+            throw new BoboException("Exception occurred when trying to retrieve driverName", e);
+        }
+    }
+
+    private PooledDataSource(String url, String username, String password) {
+        log.trace("Creating instance of {} with default pool size: {}", getClass().getSimpleName(), DEFAULT_POOL_SIZE);
+        checkArguments(url, username, password, DEFAULT_POOL_SIZE);
+        this.connectionPool = new ConcurrentLinkedQueue<>();
+        this.url = url;
+        this.username = username;
+        this.password = password;
+        this.poolSize = DEFAULT_POOL_SIZE;
+        initializePool();
+    }
+
+    private PooledDataSource(String url, String username, String password, int poolSize) {
+        checkArguments(url, username, password, poolSize);
+        log.trace("Creating instance of {} with pool size: {}", getClass().getSimpleName(), DEFAULT_POOL_SIZE);
+        this.connectionPool = new ConcurrentLinkedQueue<>();
+        this.url = url;
+        this.username = username;
+        this.password = password;
+        this.poolSize = poolSize;
+        initializePool();
+    }
+
+    private void initializePool() {
+        String driverName = Optional.ofNullable(getDriverName()).orElseThrow(() -> new IllegalArgumentException("Cannot retrieve the driver"));
+        log.trace("Initializing connection pool. Size: {}. Driver: {}", poolSize, driverName);
+        for (int i = 0; i < poolSize; i++) {
+            try {
+                var connection = new ConnectionProxy(DriverManager.getConnection(url, username, password), connectionPool);
+                connectionPool.offer(connection);
+            } catch (SQLException e) {
+                throw new BoboException("Exception occurred when trying to get connection from DriverManager", e);
+            }
+        }
+    }
+
+    private void checkArguments(String url, String username, String password, int poolSize) {
+        if (poolSize <= 0) {
+            throw new IllegalArgumentException("Pool size cannot be less than 1");
+        }
+
+        if (Objects.isNull(url) || Objects.isNull(username) || Objects.isNull(password)) {
+            throw new IllegalArgumentException(
+                    "Properties: %s, %s, %s must not be null. DataSource has not been initialized.".formatted(URL, USERNAME, PASSWORD));
+        }
+    }
+
+    private static PooledDataSource createDataSourceInstance(String url, String username, String password, int poolSize) {
+        return new PooledDataSource(url, username, password, poolSize);
+    }
+
+    private static PooledDataSource createDataSourceInstanceFromProperties() {
```

Comment:
I would create 1 more overloaded version of this method with the custom property file path.